### PR TITLE
HPCC-18250 Move gatherStats() from destructor to avoid accessing invalid memory

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman please review

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -27733,7 +27754,7 @@ class CDelayedActivityGraph : implements IActivityGraph, public CInterface
     virtual IEclGraphResults * queryLocalGraph() { throwUnexpected(); }
     virtual IRoxieServerChildGraph * queryLoopGraph() { throwUnexpected(); }
     virtual IRoxieServerActivity *queryActivity(unsigned _activityId) { return nullptr; } // MORE - may need something here!?
-
+    virtual void updateFactoryStatistics() const { }
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: override?

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -27510,7 +27526,12 @@ class CProxyActivityGraph : implements IActivityGraph, implements IThorChildGrap
     virtual IRoxieServerChildGraph * createGraphLoopInstance(IRoxieSlaveContext *ctx, unsigned loopCounter, unsigned parentExtractSize, const byte * parentExtract, const IRoxieContextLogger &logctx) { throwUnexpected(); }
     virtual const char *queryName() const { throwUnexpected(); }
     virtual IRoxieServerActivity *queryActivity(unsigned _activityId) { return nullptr; } // MORE - may need something here!?
-
+    virtual void updateFactoryStatistics() const
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: override?

File: roxie/ccd/ccdcontext.cpp [hunk 3]
```diff
@@ -1481,6 +1481,16 @@ class CRoxieContextBase : implements IRoxieSlaveContext, implements ICodeContext
 
     void cleanupGraphs()
     {
+        if (graph)
+            graph->updateFactoryStatistics();
+
+        SuperHashIteratorOf<decltype(childGraphs)::ELEMENT> iter(childGraphs);
+        ForEach(iter)
+        {
+            IActivityGraph * graph = static_cast<IActivityGraph *>(iter.query().getValue());
```
Comment (richardkchapman) [inline roxie/ccd/ccdcontext.cpp]: Confusing to name this local variable the same as a member...

File: roxie/ccd/ccdserver.cpp [hunk 4]
```diff
@@ -15777,6 +15780,8 @@ class CRoxieServerParallelGraphLoopActivity : public CRoxieServerGraphLoopActivi
         resultStream = NULL;
         resultJunction.clear();
         outputs.kill();
+        ForEachItemIn(i, iterationGraphs)
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: Bit nervous that this is not too late, given the comment on the line below?

**Comment by richardkchapman** [comment]
@ghalliday A few comments

File: roxie/ccd/ccdserver.cpp (hunk 4 above)
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp]: I agree.  I'll move it earlier.

File: roxie/ccd/ccdserver.cpp [hunk 5]
```diff
@@ -27722,24 +27745,23 @@ class CDelayedActivityGraph : implements IActivityGraph, public CInterface
...
-    virtual void abort() { throwUnexpected(); }
-    virtual void reset() { }
-    virtual void execute() { throwUnexpected(); }
-    virtual void getProbeResponse(IPropertyTree *query) { throwUnexpected(); }
-    virtual void noteException(IException *E) { throwUnexpected(); }
-    virtual void checkAbort() { throwUnexpected(); }
-    virtual IThorChildGraph * queryChildGraph() { throwUnexpected(); }
-    virtual IEclGraphResults * queryLocalGraph() { throwUnexpected(); }
-    virtual IRoxieServerChildGraph * queryLoopGraph() { throwUnexpected(); }
-    virtual IRoxieServerActivity *queryActivity(unsigned _activityId) { return nullptr; } // MORE - may need something here!?
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: note: adding override revealed this function isn't in the base interface - so deleted.

**Comment by ghalliday** [comment]
@richardkchapman addressed comments.  Please check again.

File: roxie/ccd/ccdserver.cpp [hunk 6]
```diff
@@ -27510,7 +27528,12 @@ class CProxyActivityGraph : implements IActivityGraph, implements IThorChildGrap
     virtual IRoxieServerChildGraph * createGraphLoopInstance(IRoxieSlaveContext *ctx, unsigned loopCounter, unsigned parentExtractSize, const byte * parentExtract, const IRoxieContextLogger &logctx) { throwUnexpected(); }
     virtual const char *queryName() const { throwUnexpected(); }
     virtual IRoxieServerActivity *queryActivity(unsigned _activityId) { return nullptr; } // MORE - may need something here!?
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: I suspect this should be removed too?

**Comment by richardkchapman** [comment]
@ghalliday One more comment

**Comment by ghalliday** [comment]
@richardkchapman removed function and added override to the other members of that class.

**Pull request merged**